### PR TITLE
GH-37582: [Go][Parquet] Implement Float16 logical type

### LINKED ISSUE #37582: [Go][Parquet] Implement Float16 logical type
### Describe the enhancement requested

As a follow-up to https://github.com/apache/arrow/issues/36036, we should add an equivalent implementation for Go as well.

### Component(s)

Go, Parquet

### PR DESCRIPTION
**Pull request by benibus**
### Rationale for this change

There is an active proposal for a Float16 logical type in Parquet (https://github.com/apache/parquet-format/pull/184) with C++/Python implementations in progress (https://github.com/apache/arrow/pull/36073), so we should add one for Go as well.

### What changes are included in this PR?

- [x] Adds `LogicalType` definitions and methods for `Float16`
- [x] Adds support for `Float16` column statistics and comparators
- [x] Adds support for interchange between Parquet and Arrow's half-precision float

### Are these changes tested?

Yes

### Are there any user-facing changes?

Yes

* Closes: #37582

### CODE REVIEW

**Review by benibus — COMMENTED** [review]
This should cover the base requirements, but LMK if I missed anything. There's a bit of weird special-casing involved with the column statistics since float16 requires an independent `Statistics` type despite not being physical - but maybe there's a better way.

File: go/arrow/float16/float16.go
Comment (zeroshade) [inline go/arrow/float16/float16.go:175]: Should we use this to improve the `Sign` method?

File: go/parquet/file/column_writer_types.gen.go.tmpl
Comment (zeroshade) [inline go/parquet/file/column_writer_types.gen.go.tmpl:212]: would probably be better to check `w.Descr().LogicalType()` and check if it's a Float16 column rather than make this assumption which might bite us in the future.

File: go/parquet/file/column_writer_types.gen.go.tmpl
Comment (zeroshade) [inline go/parquet/file/column_writer_types.gen.go.tmpl:233]: same as before

File: go/parquet/internal/testutils/random.go
Comment (zeroshade) [reply go/parquet/internal/testutils/random.go:379]: It might make more sense to leave the NaN's and infinities to ensure we're covering those cases for testing. Alternately we could just make sure we test those cases separately

File: go/arrow/float16/float16.go
Comment (zeroshade) [inline go/arrow/float16/float16.go:36]: might make sense to add `func Inf() float16.Num` and `func NaN() float16.Num` functions to this package

File: go/parquet/schema/reflection_test.go
Comment (zeroshade) [inline go/parquet/schema/reflection_test.go:155]: can we add a test for an optional one (instead of required) which would use `*[2]byte` as the type?
can we also add a test that uses `float16.Num` as the type?

**Review by zeroshade — CHANGES_REQUESTED** [review]
Overall this looks great, just a few nitpicks.

File: go/parquet/file/column_writer_types.gen.go.tmpl
Comment (benibus) [reply go/parquet/file/column_writer_types.gen.go.tmpl]: Updated now, but worth another look in case I misunderstood the concern.

**Comment by benibus** [comment]
Note that the updated `parquet.thrift` file that this uses is currently contained in the C++ PR https://github.com/apache/arrow/pull/36073, which will need to be synced with https://github.com/apache/parquet-format/pull/184 once it's finalized.

So, fair warning: regenerating the thrift files from this branch alone will basically break everything (for now).

File: go/parquet/file/column_writer_types.gen.go.tmpl
Comment (zeroshade) [reply go/parquet/file/column_writer_types.gen.go.tmpl]: looks good! thanks!

File: go/parquet/schema/reflection_test.go
Comment (zeroshade) [reply go/parquet/schema/reflection_test.go:155]: the expected result for the `float16.Num` would be float16 column as per the spec (fixed len binary + float16 logical type). It shouldn't need the `parquet:"logical=float16"` if the type is `float16.Num`. 

You may need to special case the code to check for the float16.Num type to allow it through

**Comment by zeroshade** [comment]
Overall this is looking good to me, though I won't merge this until after the format PR with the thrift changes gets finalized and merged.

**Comment by zeroshade** [comment]
@benibus the fix for the failing CIs has been merged, can you rebase and update this so that we can get a clean CI on it? I believe the float16 format has been merged so once you rebase and we have all green i'll be happy to merge this!

**Comment by benibus** [comment]
Rebased. A couple CI failures still, but they look unrelated.

**Comment by zeroshade** [comment]
Can you please resolve the conflicts (and potentially rebase if necessary)? We can check the CI afterwards

**Comment by benibus** [comment]
Alright, I've fixing the conflicts and rebased again. It looks like everything's green now.

**Review by zeroshade — APPROVED** [review]
LGTM

**Pull request merged**
